src/factorysetup.c

```diff
@@ -302,6 +329,114 @@ static void _api_msg(const uint8_t* input, size_t in_len, uint8_t* output, size_
     *output_len = out_len;
 }
 
+static void _free(uint8_t** buf)
+{
+    if (*buf != NULL) {
+        free(*buf);
+    }
+}
+
+static uint8_t _ble_firmware_checksum(const uint8_t* buf, size_t buf_len)
+{
+    uint8_t res = 0;
+    for (size_t i = 0; i < buf_len; ++i) {
+        res ^= buf[i];
+    }
+    return res;
+}
+
+static ble_error_code_t _verify_ble(const uint8_t* expected_ble_fw_hash, uint8_t expected_checksum)
+{
+    // We don't need to double-check the metadata, `memory_spi_get_active_ble_firmware()` and the
+    // subsequent checks cover it (the FW uses the same function to fetch the active FW).
+
+    uint8_t* __attribute__((__cleanup__(_free))) flashed_ble_fw = NULL;
+    size_t flashed_ble_fw_size = 0;
+    uint8_t flashed_ble_fw_checksum = 0;
+    if (!memory_spi_get_active_ble_firmware(
+            &flashed_ble_fw, &flashed_ble_fw_size, &flashed_ble_fw_checksum)) {
+        screen_print_debug("_setup_ble: get firmware failed", 0);
+        return BLE_ERR_READ_FW;
+    }
+    uint8_t flashed_ble_fw_hash[32] = {0};
+    if (wally_sha256(
+            flashed_ble_fw,
+            flashed_ble_fw_size,
+            flashed_ble_fw_hash,
+            sizeof(flashed_ble_fw_hash)) != WALLY_OK) {
+        Abort("_setup_ble: wally_sha256 failed");
+    }
+    if (flashed_ble_fw_size != da14531_firmware_size()) {
+        screen_print_debug("_setup_ble: size check failed", 0);
+        return BLE_ERR_FW_SIZE_MISMATCH;
+    }
+    if (flashed_ble_fw_checksum != expected_checksum) {
+        screen_print_debug("_setup_ble: checksum mismatch", 0);
+        return BLE_ERR_FW_CHECKSUM_MISMATCH;
+    }
+    if (!MEMEQ(expected_ble_fw_hash, flashed_ble_fw_hash, 32)) {
+        screen_print_debug("_setup_ble: hash check failed", 0);
+        return BLE_ERR_FW_MISMATCH;
+    }
+
+    return BLE_ERR_OK;
+}
+
+// TODO: call this as part of an API endpoint?
+static ble_error_code_t _setup_ble(void)
+{
+    if (da14531_firmware_size() > MEMORY_SPI_BLE_FIRMWARE_MAX_SIZE) {
+        return BLE_ERR_FW_TOO_LARGE;
+    }
+
+    uint8_t checksum = _ble_firmware_checksum(da14531_firmware_start(), da14531_firmware_size());
+
+    // Compute FW hash.
+    uint8_t ble_fw_hash[32] = {0};
+    if (wally_sha256(
```

Comment:
Is there a reason you don't use the `rust_sha256`?

Reply:
It's because in the same file we already use wally_sha256 :shrug: